### PR TITLE
Added Provision in request to enable cookies.

### PR DESCRIPTION
**Pull request by farisbacker**
As I understand authtoken are saved in memory, It cause issue when we have multiple container instance of guacamole server.

Can we have cookies option with request ? With stickiness enabled in load balancer. I can properly point to exact container.

Let me know, if i missed anything.

### CODE REVIEW

**Comment by pschmitt** [comment]
How could the current way of storing auth lead to issues when running in containers? I don't believe that's an issue at all. How can I trigger this? (ie. what's your setup?)

File: guacapy/client.py [hunk 1]
```diff
@@ -77,7 +83,7 @@ def __authenticate(self):
             allow_redirects=True,
         )
         r.raise_for_status()
-        return r.json()
```
Comment (pschmitt) [inline guacapy/client.py:80]: Why the change?

**Comment by farisbacker** [comment]
> How could the current way of storing auth lead to issues when running in containers? I don't believe that's an issue at all. How can I trigger this? (ie. what's your setup?)

Existing Setup :

Guacamole docker container deployed in ECS Fargate with loadbalancer(elb)  in front of it.

While invoking RESTApi we faced intermittent authentication(403) failure. AuthToken
was generated successfully, But further request with authtoken failed wih
403 error. By deep diving we understood that its because
both request was going to different container. Authtoken generated was only
valid for container in which it was invoked.

For more info:
http://apache-guacamole-general-user-mailing-list.2363388.n4.nabble.com/Is-Guacamole-RESTAPI-AuthToken-stored-in-memory-tt8966.html

I am planning is to use cookies with request and
enable stickiness in loadbalancer.

Let me know, if I understood differently.

**Comment by ToasterChicken** [comment]
To handle cookies isn't it just a matter of adding a requests.session to the Guacamole  class and change any of the requests.post to use that session?  Requests already has facilities for handling cookies in the session.

More info:
https://requests.readthedocs.io/en/master/user/advanced/

So just adding to the to the class \_\_init\_\_ :
`self.session = requests.Session()`

and changing any requests.post to self.session.post for example

Appears to work on my load balanced cluster in swarm. I don't suspect this change would impact connections otherwise with out cookies. As I understand how Requests Sessions work, this should also make the class a good citizen and maintain the connection as the script executes with some exceptions. Although in those contexts, it'll use the session + cookie for the calls until that context is closed, where it will then close the session.

I haven't tested it extensively yet, just get_connection_by_name and authenticating.

File: guacapy/client.py (hunk 1 above)
Comment (farisbacker) [reply guacapy/client.py:80]: This done to get cookies data from response. In earlier case,  I was not able to find cookies content in r.json().

**Comment by ToasterChicken** [comment]
I just tested it on a single guac node to test with sessions as the only change to support cookies and without sessions, requesting guac.get_users() 100 times running it through cProfile and this should solve this problem as well as improve the performance overall. I was doing performance testing and some simple load gen anyways on the cluster so I thought I'd make a small change to the test scripts and share.

With the current version:
601714 function calls (600300 primitive calls) in 45.237 seconds
Using Sessions:         
363108 function calls (360850 primitive calls) in 13.093 seconds

And as expected the results are the same when scaling the guacamole service to 4 instances since traefik is just going to send it all to one instance. So we're connecting once, validating the SSL certificate once, performing the SSL handshake once

I can see in the profile that SSL.py:1906(do_handshake) is called once with the session, and 100 times with out. So it's connecting once with everything that involves, validating the SSL certificate once, performing the SSL handshake once

I guess the only thing this might affect is the API Token TTL, the session will need to be restarted after it expires (default 60 minutes) assuming someone intends to use this for extremely long periods of inactivity. I guess they could put in some sort of short query as a keep alive and maybe a session retry if the particular node became unavailable and re-authenticate to another node.

**Comment by pschmitt** [comment]
Got it. Sorry for delay. LGTM.

**Comment by pschmitt** [comment]
This is included in 0.9.9: https://pypi.org/project/guacapy/0.9.9/
